### PR TITLE
Allowing non-dictionary-like bodies; Support for PUT calls

### PR DESCRIPTION
**Pull request by tsoenen**
Allowing non-dictionary body structures, additional checks on query and body parameter config

### CODE REVIEW

File: qrest/conf.py [hunk 1]
```diff
@@ -223,8 +223,8 @@ def validate(self):
             self.headers = {}
 
         # method  --------------------
-        if self.method not in ["GET", "POST"]:
-            raise RestClientConfigurationError("method must be GET or POST")
+        if self.method not in ["GET", "POST", "PUT"]:
+            raise RestClientConfigurationError("method must be GET, POST or PUT")
```
Comment (swinkels-at-basf) [inline qrest/conf.py:227]: ;) Needed to allow non-dictionary like bodies?

**Review by swinkels-at-basf — CHANGES_REQUESTED** [review]

**Comment by swinkels-at-basf** [comment]
I should not forget, thanks for the pull request!

File: docs/source/configuration.rst [hunk 2]
```diff
@@ -294,6 +294,11 @@ QueryParameter for key ``choices_param``, whose 'remote' name is
 
 will request the resource at URL http://example.com?ChoicesParam=value
 
+For a BodyParameter, it is allowed that the name attribute has value None. In that
```
Comment (swinkels-at-basf) [inline docs/source/configuration.rst:297]: The following is **not** something you have to address as part of this PR **so consider it FYI**

* This update to the documentation makes me realize something, if you can supply `name=None`, should that make `name` an optional parameter?
* So to provide a custom, non-dictionary body, the user has to provide `name=None`. That doesn't seem to be an obvious, or telling, way to instruct the use of a non-dictionary body. What about creating a new subclass of `BodyParameter` that doesn't have a `name` parameter (and internally uses `None` as the value of `name`), e.g. `AnybodyParameter` - I'm just kidding about the name, but the suggestion stands ;).

**Review by swinkels-at-basf — CHANGES_REQUESTED** [review]
Just one comment to address, viz. the one about the missing mention to PUT in `configuration.rst`.

The other comment is a FYI. Do feel free to pick it up but even if you do, it doesn't have to be done as part of this PR.

File: docs/source/configuration.rst [hunk 3]
```diff
@@ -294,6 +294,11 @@ QueryParameter for key ``choices_param``, whose 'remote' name is
...
+case, the value that is passed to the BodyParameter will be added as such to the body
+of the request, not as a key/value pair. This is useful for bodies with a
+non-dictionary structure. In that case, a resource can have only one BodyParameter.
+For QueryParameters, a name attribute with value None is not allowed.
```
[the remark this answers is not in the document]
Comment (tsoenen) [reply docs/source/configuration.rst:300]: Done

File: docs/source/configuration.rst (hunk 2 above)
Comment (tsoenen) [reply docs/source/configuration.rst:296]: I agree that this approach doesn't clearly capture what a `name=None` enables. I see it as a first step to give the qrest user the ability to wrap endpoints with more complex body structures (e.g. non-dictionary and nested dictionaries (which was already possible, but only the root keys can be added to a resource as BodyParameter)). 

In a second phase, I'd like to improve the BodyParameter to support json schemas. The body that the enduser provides can be validated with this schema, and the schema can be used to help the end user building the body, e.g. by making a template available for them to fill.

**Review by swinkels-at-basf — APPROVED** [review]

**Pull request merged**